infrastructure/ssz/src/main/java/tech/pegasys/teku/infrastructure/ssz/tree/SimpleLeafNode.java

```diff
@@ -22,25 +22,39 @@
 
 class SimpleLeafNode implements LeafNode, TreeNode {
 
-  private final byte[] data;
+  private final Bytes data;
+  private Bytes32 hashTreeRoot;
 
   public SimpleLeafNode(Bytes data) {
     checkArgument(data.size() <= MAX_BYTE_SIZE);
-    this.data = data.toArrayUnsafe();
+    if (data.size() == MAX_BYTE_SIZE) {
+      // if data is Bytes32, it will pass throw with no object creation
+      // otherwise translate types to Bytes32
+      this.hashTreeRoot = Bytes32.wrap(data.copy());
+      this.data = hashTreeRoot;
+      return;
+    }
+    // If we store data as is, some Bytes instances (ie ConcatenatedBytes) will
+    // perform worse during serialization. So we want to translate them to a
+    // single array-backed Bytes
+    //
+    // in this case, the following seems to perform better than Bytes.wrap(data.copy())
+    this.data = Bytes.wrap(data.toArrayUnsafe());
   }
 
   @Override
   public Bytes getData() {
-    return Bytes.wrap(data);
+    return data;
   }
 
   @Override
   public Bytes32 hashTreeRoot() {
-    if (data.length == MAX_BYTE_SIZE) {
-      return Bytes32.wrap(data);
-    } else {
-      return Bytes32.wrap(Arrays.copyOf(data, MAX_BYTE_SIZE));
+    if (hashTreeRoot != null) {
+      return hashTreeRoot;
     }
```

Comment:
this seems like as big an improvement as the rest of it, actually storing hash and not constantly rebuilding...

Reply:
yes that's the main idea. Which is a 100% win in the Bytes32 case, where hashTreeRoot and data are actually the same thing.
In case of smaller data size is where we pay the bill, lower GC pressure but higher mem consumption.